server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.stats;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiConsumer;
+
+import static org.opensearch.common.settings.Setting.Property.NodeScope;
+
+/**
+ * A class caches use to internally keep track of their stats across multiple dimensions. Not intended to be exposed outside the cache.
+ */
+public class StatsHolder implements Writeable {
+    /**
+     * For memory purposes, don't track stats for more than this many distinct combinations of dimension values.
+     */
+    public static final Setting<Integer> MAX_DIMENSION_VALUES_SETTING = Setting.intSetting("cache.stats.max_dimension", 20_000, NodeScope);
+
+    // The list of permitted dimensions.
+    private final List<String> dimensionNames;
+
+    /**
+     * Determines which combinations of dimension values are tracked separately by this StatsHolder. In every case,
+     * incoming keys still must have all dimension values populated.
+     */
+    public enum TrackingMode {
```

Comment:
This was meant to be up to the cache. 

I originally had just the equivalent of ALL_COMBINATIONS but Sagar and I thought it might be good to let whatever cache uses this specify in advance what they want to aggregate. I think it would also make sense to just drop the modes and keep the ALL_COMBINATIONS logic, though.